mlaut/highlevel/strategies.py

```diff
@@ -0,0 +1,418 @@
+import pandas as pd
+import inspect
+from mlaut.shared.static_variables import HDF5_EXTENTION
+from tensorflow.python.keras.models import load_model
+from sklearn.preprocessing import OneHotEncoder
+import numpy as np
+import warnings
+
+# REGRESSOR_TYPES = BaseRegressor
+# CLASSIFIER_TYPES = BaseClassifier
+# ESTIMATOR_TYPES = [REGRESSOR_TYPES, CLASSIFIER_TYPES]
+
+
+class BaseStrategy(BaseEstimator):
+    """
+    Abstract base strategy class.
+
+    Implements attributes and operations shared by all strategies,
+    including input and compatibility checks between passed estimator,
+    data and task.
+    """
+
+    def __init__(self, estimator, check_input):
+        self._estimator = estimator
+        self.check_input = check_input
+        self._task = None
+        self._check_estimator_compatibility(estimator)
+
+
+    @property
+    def name(self):
+        """Makes attribute accessible, but read-only.
+        """
+        return self._name
+
+    @property
+    def estimator(self):
+        """Makes attribute accessible, but read-only.
+        """
+        return self._estimator
+
+    # def __getitem__(self, key):
+    #     """
+    #     Provide read only access via keys to the private traits
+    #     """
+    #     if key not in self._traits.keys():
+    #         raise KeyError
+    #     return self._traits[key]
+
+    def fit(self, task, data):
+        """
+        Fit the strategy to the given task and data.
+
+        Parameters
+        ----------
+        task : Task
+            Task encapsualting metadata information on feature and target variables to which to fit the data to.
+        data : pandas.DataFrame
+            Dataframe with feature and target variables as specified in task.
+
+        Returns
+        -------
+        self : an instance of the self
+        """
+        if self.check_input:
+            self._validate_data(data)
+
+        # Check task compatibility with strategy
+        self._check_task_compatibility(task)
+        self._task = task
+
+        # Set metadata if not already set
+        if self._task.metadata is None:
+            self._task.set_metadata(data)
+
+        # strategy-specific implementation
+        return self._fit(data)
+
+    def _check_task_compatibility(self, task):
+        """
+        Check compatibility of task with strategy
+        """
+        # TODO replace by task-strategy compatibility lookup registry
+        if hasattr(task, '_case'):
+            if self._case != task._case:
+                raise ValueError("Strategy <-> task mismatch: The chosen strategy is incompatible with the given task")
+        else:
+            raise AttributeError("The passed case of the task is unknown")
+
+    def _check_estimator_compatibility(self, estimator):
+        """
+        Check compatibility of estimator with strategy
+        """
+        # TODO perform checks for passed estimator to ensure that it is compatible with the Orchestrator
+        pass
```

Comment:
I'd remove this for now when it's not implemented and put it on the dev roadmap, calling a function which is supposed to check something but does nothing may be dangerous and we may forget about it

Reply:
ok